Rust/src/common/divector.rs

```diff
@@ -0,0 +1,174 @@
+            low: vec![0.0;dimension]
+        }
+    }
+
+    pub fn new(high : &[f64], low :&[f64]) -> Self{
+        assert!(high.len() == low.len(), " incorrect lengths");
+        DiVector{
+            high : Vec::from(high),
+            low : Vec::from(low)
+        }
+    }
+
+    pub fn assign_as_probability_of_cut(&mut self, bounding_box: &BoundingBox, point: &[f32]) {
+        let minsum: f64 = self.low
+            .iter_mut()
+            .zip(bounding_box.get_min_values())
+            .zip(point)
+            .map(|((x, &y), &z)| if y - z > 0.0 {
+                *x = (y - z) as f64;
+                *x
+            } else {
+                *x = 0.0;
+                *x
+            })
+            .sum();
+        let maxsum: f64 = self.high
+            .iter_mut()
+            .zip(point)
+            .zip(bounding_box.get_max_values())
+            .map(|((x, &y), &z)| if y - z > 0.0 {
+                *x = (y - z) as f64;
+                *x
+            } else {
+                *x = 0.0;
+                *x
+            })
+            .sum();
+
+        let sum = minsum + maxsum;
+        if sum != 0.0 {
+            self.scale(1.0/(bounding_box.get_range_sum() + sum));
+        }
+    }
+
+    pub fn assign_as_probability_of_cut_with_missing_coordinates(
+        &mut self,
+        bounding_box:&BoundingBox,
+        point: &[f32],
+        missing_coordinates: &[bool]
+    ) {
+        let minsum: f64 = self.low
+            .iter_mut()
+            .zip(bounding_box.get_min_values())
+            .zip(point)
+            .zip(missing_coordinates)
+            .map(|(((x, &y), &z), &b)| if !b && y - z > 0.0 {
+                *x = (y - z) as f64;
+                *x
+            } else {
+                *x = 0.0;
+                *x
+            })
+            .sum();
+        let maxsum: f64 = self.high
+            .iter_mut()
+            .zip(point)
+            .zip(bounding_box.get_max_values())
+            .zip(missing_coordinates)
+            .map(|(((x, &y), &z), &b)| if !b && y - z > 0.0 {
+                *x = (y - z) as f64;
+                *x
+            } else {
+                *x = 0.0;
+                *x
+            })
+            .sum();
+
+        let sum = minsum + maxsum;
+        if sum != 0.0 {
+            self.scale(1.0/(bounding_box.get_range_sum() + sum));
+        }
+    }
+
+
+    pub fn assign(&mut self, other :&DiVector){
+        for (x, &y) in self.high.iter_mut().zip(&other.high) {
+            *x = y;
+        }
+        for (x, &y) in self.low.iter_mut().zip(&other.low) {
+            *x = y;
+        }
+    }
+
+    pub fn add_from(&mut self, other: &DiVector, factor : f64) {
+
+        for (x, &y) in self.high.iter_mut().zip(&other.high) {
+            *x += y * factor;
+        }
+        for (x, &y) in self.low.iter_mut().zip(&other.low) {
+            *x += y * factor;
+        }
+
+    }
+
+
+    pub fn add_to(&self, other: &mut DiVector) {
```

Comment:
Couldn't you implement this as:
```
other.add_from(self)
```

Reply:
Yes ;) But we need an explicit function that given collection of R's in R_1,R_2... iterates over and collects to a type S. In this case R = S; but need not be (think of approximate median of a sequence of values). Also, even though we wish to iterate, the type of the iterator would be different based on rayon par_iter() or native iter(). This will probably get rewritten once the TAR collector framework is implemented,